File: app/src/main/java/kashyap/in/androidbestpractices/data/CustomConverters.java
package kashyap.in.androidbestpractices.data;

import androidx.room.TypeConverter;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class CustomConverters {
    @TypeConverter
    public static ArrayList<String> fromString(String value) {
        Type listType = new TypeToken<ArrayList<String>>() {
        }.getType();
        return new Gson().fromJson(value, listType);
    }

    @TypeConverter
    public static List<Integer> fromStringInteger(String value) {
        Type listType = new TypeToken<ArrayList<Integer>>() {
        }.getType();
        return new Gson().fromJson(value, listType);
    }

    @TypeConverter
    public static String fromArrayListOfInteger(List<Integer> list) {
        Gson gson = new Gson();
        return gson.toJson(list);
    }

    @TypeConverter
    public static String fromArrayListOfString(ArrayList<String> list) {
        Gson gson = new Gson();
        return gson.toJson(list);
    }

    @TypeConverter
    public static String fromArrayOfInt(int[] list) {
        Gson gson = new Gson();
        return gson.toJson(list);
    }

    @TypeConverter
    public static int[] fromIntArrayString(String value) {
        Type listType = new TypeToken<int[]>() {
        }.getType();
        return new Gson().fromJson(value, listType);
    }
}

File: app/src/main/java/kashyap/in/androidbestpractices/data/CustomConverters.kt
package kashyap.`in`.androidbestpractices.data

import androidx.room.TypeConverter
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.util.*

class CustomConverters {
    @TypeConverter
    fun fromString(value: String?): ArrayList<String> {
        val listType =
            object : TypeToken<ArrayList<String?>?>() {}.type
        return Gson().fromJson(value, listType)
    }

    @TypeConverter
    fun fromStringInteger(value: String?): List<Int> {
        val listType =
            object : TypeToken<ArrayList<Int?>?>() {}.type
        return Gson().fromJson(value, listType)
    }

    @TypeConverter
    fun fromArrayListOfInteger(list: List<Int?>?): String {
        return Gson().toJson(list)
    }

    @TypeConverter
    fun fromArrayListOfString(list: ArrayList<String?>?): String {
        return Gson().toJson(list)
    }

    @TypeConverter
    fun fromArrayOfInt(list: IntArray?): String {
        return Gson().toJson(list)
    }

    @TypeConverter
    fun fromIntArrayString(value: String?): IntArray {
        val listType = object : TypeToken<IntArray?>() {}.type
        return Gson().fromJson(value, listType)
    }
}